### DISCUSSION TITLE
Elsa 3: How to get the request body of an HttpEndpoint?

### DISCUSSION THREAD

**Discussion by ploef** [elsa-workflows/elsa-core#4030] [Q&A]
Hi,

Can anybody help me out how to process the body received with the HttpEndpoint?
I'm trying to post Json to a HttpEndpoint, that seems to go ok. I'm setting the "ParsedContent" property to an variable:

<img width="918" alt="image" src="https://github.com/elsa-workflows/elsa-core/assets/1884192/ba0ab934-405d-4a4d-8ff5-29389375d47a">

Is that the correct way to do this? And how to handle the body in the next activity (I have to do some deserialization and mapping)?

**Comment by sfmskywalker** [comment]
That's the right way to do it 👍🏻 

I just created a [sample project](https://github.com/elsa-workflows/elsa-core/blob/v3/src/samples/aspnet/Elsa.Samples.AspNet.HttpEndpoints/Workflows/SubmissionWorkflow.cs) based on your question to demonstrate using the variable in a subsequent activity.

The workflow looks like this:

```csharp
/// <summary>
/// A workflow that acts as an HTTP endpoint to which a client can POST a JSON payload.
/// </summary>
public class SubmissionWorkflow : WorkflowBase
{
    protected override void Build(IWorkflowBuilder builder)
    {
        // The payload will be deserialized into a dynamic object. The JSON might look like this:
        // {
        //   "name": "John Doe",
        //   "email": "john.doe@localhost"
        // }
        var payloadVariable = builder.WithVariable<dynamic>().WithMemoryStorage();

        builder.Root = new Sequence
        {
            Activities =
            {
                new HttpEndpoint
                {
                    Path = new("submit"),
                    SupportedMethods = new(new[]{HttpMethods.Post}),
                    CanStartWorkflow = true,
                    ParsedContent = new(payloadVariable)
                },
                new WriteHttpResponse
                {
                    ContentType = new(MediaTypeNames.Application.Json),
                    Content = new(context =>
                    {
                        var payload = payloadVariable.Get(context)!;
                        var name = payload.name;
                        var email = payload.email;
                        
                        return new
                        {
                            Message = $"Dear {name}, thank you for submitting your information! We'll send you an email at {email} shortly.",
                        };
                    }),
                }
            }
        };
    }
}
```

The key is to simply call the `Get` method on your workflow variable, passing in the current activity execution context.
Every `Input` property of any activity provides access to this context, which means you can use workflow variables for every activity.

**Reply by ploef** [reply]
Thanks @sfmskywalker that works like a charm!
If I would like to take it one step further and want to use the actual class representing the request body, how do I pass that to the next Activity? At the moment I'm getting the following error message:

![image](https://github.com/elsa-workflows/elsa-core/assets/1884192/a740cdf8-fb0f-4c26-8a76-b5a743711444)

![image](https://github.com/elsa-workflows/elsa-core/assets/1884192/f5f8268a-1940-489f-9e11-d76b7b73418c)

![image](https://github.com/elsa-workflows/elsa-core/assets/1884192/484ca053-3146-4858-9842-0da0ea06697d)

**Reply by sfmskywalker** [reply]
As it turns out, you ran into a little bug there that happened tome as well one or two days ago. I have fixed it since, and to make sure it also fixes your scenario, I updated my earlier sample project. If you update to the latest **preview** version (not the rc1 from nuget), it should work. If not, please do let me know. Thanks!
